### ISSUE TITLE
Make delete of Theano function delete all of it in 1 call to gc.collect()

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#4275]
https://groups.google.com/d/topic/theano-users/A52Y64VaN-I/discussion

It happened a few times that user think there is a link in a compiled Theano function, but the problems is circular reference in our representation that request multiple the cycle detection pass of Python gc.

What about implementing FunctionGraph.\__del__  and deleting the clients fields to break some the cycles. After this one, we should test again to see if there is other cycles.

Is there inconvenience of implementing \__del__ to break manually the cycles?

**Comment by nouiz** [comment]
https://docs.python.org/2/reference/datamodel.html#object.__del__ recommand to not implement \__del__

The cycle in Theano graph cause many problems, at least this one and make it more complicated to fix deepcopy max recursion.

So what about changing the way we represent clients? What about making a dict in FunctionGraph whose key is an Apply node while the value is the clients list? To not break the inferface, we could make Variable.clients, be a properties that check self.fgraph.clients_dict and return it. (This would be slower, so we could do that with a warning to help people change there code thta use that outside Theano).

**Comment by Sentient07** [comment]
@nouiz The former moment was what i have written in my proposal, a way to break recursion in input-output. I understand the alternative method in your second comment. But could you please elaborate more on the warning ?

**Comment by nouiz** [comment]
If you mean "to not use \__del__" as the warning, it mean, my first suggestion isn't a good one. We should not use \__del__.

**Comment by Sentient07** [comment]
But using a dict would make it slow ? I had thought the warning was related to the task getting slow. I will update the latter with the former in the proposal. My only worry is it might be slow..

**Comment by nouiz** [comment]
The problem isn't the slowness. The problem is a a combination of how thinks work in Theano and how Python garbage collector work. Python GC free object when the refcount reach 0. It detect cycle of reference from time to time, but this isn't the most efficient way and delay the clean up to when Python this GC cycle detection. This produce conter intuitive result in some cases and make Theano user think there is a memory leak in Theano. Which isn't the case.

**Comment by Sentient07** [comment]
Yes, now it makes sense to me why we will resort to the second method. Could you please explain more about this ? Is that slower part about the delaying the cleanup?
> (This would be slower, so we could do that with a warning to help people change there code thta use that outside Theano

**Comment by nouiz** [comment]
There is 2 different problems:

1) the max recursion
2) the delayed python GC of compiled Theano function.

Both of them are related to how Theano have its internal representation. 2) is only caused by the internal representation having circular reference. To fix one 1) We need to break circular reference and after use a good order of operation.

1) can be fixed independently of 2). But if we fix 2) first, this will make the fix for 1) faster as part of it will be done.

Fixing both 1 and 2 is more difficult, but fix more problems. Currently, for GSoC, we only thinked about 1). I didn't talk with other Theano dev about my proposal to fix 2). It have some inconvenience, but if we fix this, it could help many recursion problems.

@abergeron @lamblin when you have the time, what do you think of the proposed fix to change the representation of clients to not cause cycle in the internal representation?

**Comment by Sentient07** [comment]
Thanks for this explanation. I understand now. As you have asked me earlier, I have allotted 2 weeks for this (and little work WRT pickling). So I think I would have time for both 1) and 2). If you think it would take more than that, I will start the work little earlier depending on what @abergeron and @lamblin say
